### ISSUE TITLE
Snapshot - lstat errors in ignored directories/files

### ISSUE THREAD

**Issue by stnokott** [kopia/kopia#3334]
Hey there,

**TLDR: is it intended that ignored folders are still checked via `lstat` when snapshotting/estimating?**

---

I have a folder `/mnt/appdata/seafile/fuse` which I want Kopia to ignore, since I know _it won't be able to access it_.
In my policy for `/mnt/appdata`, I have ignored `**/seafile/fuse`.

Output of `kopia policy get /mnt/appdata`:
```
Policy for root@kopia:/mnt/appdata:

[...]

Files policy:
  Ignore cache directories:          true   inherited from (global)
  Ignore rules:                             (defined for this target)
    **/seafile/fuse
    **/kopia/cache
    **/AI_RecycleBin/**
    @*
  Read ignore rules from files:             inherited from (global)
    .kopiaignore
  Scan one filesystem only:          true   inherited from (global)

[...]
```

---

When running `kopia snapshot estimate --log-level=debug /mnt/appdata`, I get the following output:
```
Analyzing...
[...]
Analyzing seafile...
ERROR Error in seafile: error reading directory: lstat /mnt/appdata/seafile/fuse: transport endpoint is not connected
[...]
Snapshot includes 103007 file(s), total size 13 GB
   100 MB...  1 GB:       5 files, total size 737.6 MB
   10 MB ...100 MB:      97 files, total size 3 GB
   1 MB  ... 10 MB:     961 files, total size 2 GB
   100 KB...  1 MB:   18568 files, total size 4.7 GB
   10 KB ...100 KB:   60530 files, total size 2.5 GB
   1 KB  ... 10 KB:   19647 files, total size 86.7 MB
   0 B   ...  1 KB:    3199 files, total size 1.4 MB

Snapshot excludes no files.
Snapshot excludes 3 directories. Examples:
 - @Recycle
 - @Recently-Snapshot
 - kopia/cache
Encountered 1 error(s).

Estimated upload time: 2h53m29s at 10 Mbit/s
```
(see how `kopia/cache` was correctly ignored)

Output of `kopia snapshot create --log-level=debug --log-dir-detail=10 /mnt/appdata`:
```
[...]
Snapshotting root@kopia:/mnt/appdata ...
DEBUG uploading {"source":"root@kopia:/mnt/appdata","previousManifests":0,"parallel":4}
DEBUG running action before-snapshot-root on /mnt/appdata policy.ActionCommand{Command:"", Arguments:[]string(nil), Script:"/app/scripts/before_snapshot.sh", TimeoutSeconds:300, Mode:""}
DEBUG snapshotted directory     {"path":".","error":"error reading directory: lstat /mnt/appdata/seafile/fuse: transport endpoint is not connected","dur":"168.656µs"}
DEBUG [STORAGE] GetBlob {"blobID":"kopia.maintenance","offset":0,"length":-1,"outputLength":3314,"error":null,"duration":"120.772806ms"}
DEBUG not due for full maintenance cycle until 2023-09-20 14:51:13.089553458 +0200 CEST
DEBUG not due for quick maintenance cycle until 2023-09-20 11:51:57.136634372 +0200 CEST
DEBUG not due for maintenance
DEBUG [STORAGE] PutBlob {"blobID":"_log_20230920113749_7a7b_1695202669_1695202671_1_fb9c77af308b6b37276f5eb4278f908d","length":2118,"error":null,"duration":"50.494599ms"}
DEBUG closing active connection.
DEBUG [STORAGE] Close   {"error":null,"duration":"11.95757ms"}
ERROR upload error: transport endpoint is not connected
```

The error is expected, that is why I excluded the directory in the first place.

**So why does Kopia still check the directory when I explicitly ignored it?**
Is there a way to force Kopia to fully ignore it?

I am aware of the possibility of using `--ignore-dir-errors`, but that feels more like a workaround and would possibly ignore other, valid, errors, so I would want to avoid using it.

Perhaps I'm misunderstanding how the ignore patterns work, but seeing that the ignore rule for `kopia/cache` with the exact same ignore pattern worked, makes me think that that's not the root issue. (I also tried with multiple other ignore patterns, `fuse`, `**/fuse/**`, `seafile/fuse` etc.)

Thanks for helping out!

**Comment by lupusA** [comment]
Hi @stnokott ,

The cache might be excluded correctly because it has a cachedir tag inside it. The individual excluded folders through policies do not have these tags.

I guess the estimation etc. has to respect the policy correctly which it currently does not as it seems. 

This needs to be checked.

Cheers

**Comment by stnokott** [comment]
@lupusA 
> The cache might be excluded correctly because it has a cachedir tag inside it.

This doesn't seem to be the case.
After setting `--ignore-cache-dirs` to `false` for the policy `/mnt/appdata`, the `kopia/cache` directory is still ignored.
Then removing `**/kopia/cache` from the ignore patterns results in the directory not being ignored, so it's definitely the ignore pattern that is at work here.

**Comment by lupusA** [comment]
Can you check:

https://kopia.io/docs/advanced/kopiaignore/#ignoring-files-and-folders-in-snapshots

I can check the ignore patterns in the evening.

Cheers

**Comment by stnokott** [comment]
Yeah I checked there when trying multiple ignore patterns.

I think the pattern is not the issue, I can exclude directories and files which Kopia can read just fine.

I think this is the root problem:

## Current behaviour
1) Iterate through files
2) Find folder
3) Run `lstat` on folder
4) Throw an error for directories it cannot `lstat` (`/mnt/appdata/seafile/fuse`)
5) Check if folder should be ignored
  5.1. Skip folder if it should be ignored
(ignore rules are only checked after `lstat`ing)

## Suggested behaviour
1) Iterate through files
2) Find folder
3) Check if folder should be ignored
  3.1. Skip folder if it should be ignored
4) Run `lstat` on folder
(ignore rules are checked before `lstat`ing)

---

It's possible I'm misunderstanding the flow here, this is my high-level understanding.

**Comment by lupusA** [comment]
Hi @stnokott, 

can you ping me on slack? I've created a folder structure and estimating the folder size. I tried to rebuild your scenario by removing the read and write access to a specific folder within my snapshot target. 

The estimation ran into failures reading the folder. After setting the policy it estimated just fine. 


[out.txt](https://github.com/kopia/kopia/files/12675856/out.txt)

Cheers,

**Comment by lupusA** [comment]
Is seafile/fuse mounted? It seems to be a fuse filesystem.

**Comment by stnokott** [comment]
@lupusA 

Thanks for taking the time to investigate!

Perhaps I need to clarify a few things:
- I usually use KopiaUI for managing policies etc, but I always validate by running `kopia policy get [...]`
- As such, I never used a `.kopiaignore` file, just the UI and CLI

That being said, I ran the following commands like you requested, amending deleting the policy instead of deleting the `.kopiaignore` file:
```bash
root@kopia:/app# kopia policy delete /mnt/appdata
Removing policy on "root@kopia:/mnt/appdata"...

root@kopia:/app# kopia policy set root@kopia:/mnt/appdata --add-ignore="/seafile/*"
Setting policy for root@kopia:/mnt/appdata
 - adding "/seafile/*" to "ignore rules"

root@kopia:/app# kopia policy set /mnt/appdata --add-ignore="/seafile/*"
Setting policy for root@kopia:/mnt/appdata
 - adding "/seafile/*" to "ignore rules"
 
root@kopia:/app# kopia policy get /mnt/appdata
Policy for root@kopia:/mnt/appdata:

Retention:
  Annual snapshots:                     1   inherited from (global)
  Monthly snapshots:                    2   inherited from (global)
  Weekly snapshots:                     0   inherited from (global)
  Daily snapshots:                      0   inherited from (global)
  Hourly snapshots:                     0   inherited from (global)
  Latest snapshots:                     3   inherited from (global)
  Ignore identical snapshots:       false   inherited from (global)

Files policy:
  Ignore cache directories:          true   inherited from (global)
  Ignore rules:                             (defined for this target)
    /seafile/*
  Read ignore rules from files:             inherited from (global)
    .kopiaignore
  Scan one filesystem only:         false   inherited from (global)

Error handling policy:
  Ignore file read errors:          false   inherited from (global)
  Ignore directory read errors:     false   inherited from (global)
  Ignore unknown types:              true   inherited from (global)

Scheduling policy:
  Scheduled snapshots:
    None.
  Manual snapshot:                  false   inherited from (global)

Uploads:
  Max parallel snapshots (server/UI):   1   inherited from (global)
  Max parallel file reads:              -   inherited from (global)
  Parallel upload above size:      2.1 GB   inherited from (global)

Compression disabled.

No actions defined.

Logging details (0-none, 10-maximum):
  Directory snapshotted:                5   inherited from (global)
  Directory ignored:                    5   inherited from (global)
  Entry snapshotted:                    0   inherited from (global)
  Entry ignored:                        5   inherited from (global)
  Entry cache hit:                      0   inherited from (global)
  Entry cache miss:                     0   inherited from (global)
```

When now running `kopia snapshot estimate /mnt/appdata`, I unfortunately get the same error...
Not sure what else to do here.

I understand it is difficult to reproduce the issue on your end, I will help in any way I can if possible.

**Comment by lupusA** [comment]
Let me try to recreate it over the weekend. I need to setup fuse first :-)

**Comment by stnokott** [comment]
I've done some digging in the Kopia codebase and perhaps I found something.
First, let me draw the call graph of how we get from snapshot estimation to an error from `lstat`.

---

We begin with the `estimate` function which gets a new `ignorefs` instance from `Estimate` and recursively calls `IterateEntries()` on all subdirectories:
https://github.com/kopia/kopia/blob/062e3277f6526c97776d65b57c1e06632dc55e05/snapshot/snapshotfs/estimate.go#L133-L141

`IterateEntries()` in `ignorefs` in turn calls `IterateEntries()` on the directory:
https://github.com/kopia/kopia/blob/062e3277f6526c97776d65b57c1e06632dc55e05/fs/ignorefs/ignorefs.go#L150-L168

That directory is an instance of `filesystemDirectory`. Its `IterateEntries()` function reads the contents of the directory and calls `toDirEntryOrNil()` on each one:
https://github.com/kopia/kopia/blob/062e3277f6526c97776d65b57c1e06632dc55e05/fs/localfs/local_fs.go#L147-L180

Now, `toDirEntryOrNil()` contains the `lstat` call which produces the error I get:
https://github.com/kopia/kopia/blob/062e3277f6526c97776d65b57c1e06632dc55e05/fs/localfs/local_fs.go#L134-L145

Now, let's check where the ignore patterns are checked.
This happens in `IterateEntries()` in `ignorefs`, our second step in the call graph, when calling `maybeWrappedChildEntry()` in [line 162](https://github.com/kopia/kopia/blob/062e3277f6526c97776d65b57c1e06632dc55e05/fs/ignorefs/ignorefs.go#L162).
In other words, the ignore patterns are checked in the callback of `filesystemDirectory.IterateEntries()`, but that only gets called _after_ our call to `lstat`.

To summarize, we call `lstat` on each entry in our folder (in a subcall to `IterateEntries`) _before_ checking if it should be ignored.
In most cases, this is probably fine since `lstat` does not require read permissions to return successful (which is probably why you weren't able to reproduce the issue with file permissions only). Oddly enough though, `lstat` doesn't seem to like the way I have set up my FUSE mount.

Let's be clear, I am aware that my specific setup is very likely a rare edge case, but I feel like such cases where `lstat` fails on ignored entries can be avoided by checking first if an entry should be processed and only then running I/O operations on it.

After all, `os.ReadDir()` on the parent directory yields all the information we need to be able to determine if a sub-entry should be ignored.

TLDR: it's probably a combination of `lstat` not liking unmounted FUSE directories and Kopia running `lstat` regardless of whether the entry should be ignored.

---

Sorry for this wall of text and thanks for being this kind and patient with me, I'm not good at expressing my thoughts...
If something is unclear in my chain of thinking, please ask!

**Comment by jkowalski** [comment]
This is great investigation. Fixing this issue would be great  but is quite hard due to the layering situation (where ignorefs simply wraps any other filesystem, which in general case can also be repofs during migration). I'm open to suggestions, but I don't think we can avoid major refactoring of `localfs` (which is probably needed anywa for other reasons, such as performance).

**Comment by LakeishaKowalczyk** [comment]
> which is probably needed anywa for other reasons, such as performance

This is quite regrettable, perhaps it would be possible to allow settings to ignore specific errors, or to ignore errors from files specified as ‘ignore’ in the policy

**Comment by stnokott** [comment]
Perhaps the `Directory` interface's `IterateEntries` method could be changed so that instead of returning an error (and thus immediately ending any iteration), it calls the callback with the error.

So instead of this:
```go
IterateEntries(ctx context.Context, cb func(context.Context, Entry) error) error
```
I suggest something like this:
```go
IterateEntries(ctx context.Context, cb func(context.Context, Entry, error) error) error
```

This would allow the caller of `IterateEntries` to decide what to do when an error happens instead of the callee ending the iteration.
With this change, `localfs` could decide to ignore the error when one occurs on an ignored entry.

While this adds complexity to the `Directory` interface, it increases flexibility and reduces the amount of required refactoring.
Please let me know what you think.

**Comment by jfburdet** [comment]
For info, i'm having the same problem, with https://keybase.io/ fuse client. I can't exclude it and my snapshots are failing : 

```
kopia snapshot estimate / --log-level=debug
[...]
ERROR Ignored error in home/jfburdet: error reading directory: lstat //home/jfburdet/keybase: permission denied
ERROR Ignored error in home: error reading directory: lstat //home/jfburdet/keybase: permission denied
ERROR Ignored error in .: error reading directory: lstat //home/jfburdet/keybase: permission denied
DEBUG [STORAGE] Close	{"error":null,"duration":"381ns"}
ERROR error estimating: error reading directory: lstat //home/jfburdet/keybase: permission denied
```

**Comment by jfburdet** [comment]
Bumping this again. First of all, forgive me if the following contribution sounds stupid, but the Go ecosystem is totally unknown to me.

I was able to fix my problem by changing kopia/fs/localfs/local_fs.go/toDirEntryOrNil function to this. It seems to me that it make sense and dosn't break the implicit function contract given by its name : If an entry somehow could not be read, it can/should return nil, isnt'it ? Anyway, it fixes my problem but break some end_to_end tests ...

```go
func toDirEntryOrNil(dirEntry os.DirEntry, prefix string) fs.Entry {
	fi, err := os.Lstat(prefix + dirEntry.Name())
	if err != nil {
		return nil
	}

	return entryFromDirEntry(fi, prefix)
}
```

Before going further trying to fix tests and submit a PR, I wanted your input about thix "dumb" fix : Is it a valid way to thing about this problem ?

**Comment by iGeorgeX** [comment]
I have the same problem with a fuse filesystem. I try to exclude it, but kopia still does an lstat before it tries to ignore the directory and the snapshots fail. Is there a workaround?

**Comment by iGeorgeX** [comment]
The problem still exists in version v0.18.1.

**Comment by clueo8** [comment]
I'm on v0.19.0, stumbled upon this issue as well with a FUSE mount, can't get past it:

ignored error in .: error reading directory: lstat /home/<user>/.gvfs: permission denied
